クライアント側のコードからRPCやWebSocketを極力隔離させたい

kmurayama-aim opened the pull request:
## 目標
・WebSocketの設定を行う専用のクラスを用意したい
・全体的にRPCのことを知りすぎているので、引き離す
・なるべくWebSocketとMainControllerを引き離す

## 理由
アプリケーション層とドメイン層を極力引き離したい

## 技術的変更点概要
・MainControllerが行っていたWebSocketへのイベント追加をWebSocketInitializerに任せた
・サーバーからのメッセージをクライアントのクラスでラップして使うようにした
・サーバーへのメッセージを送る際、MainControllerがJsonやWebSocketの情報を知らずに済むようにした
・Jsonのシリアライズとデシリアライズを行うクラスを用意した

## ToDo
完了

## 今回保留した項目
・全体的にまだ役割を持ちすぎている
・MessageSenderToServerクラスはJsonとWebSocketのことを知らなくても動くようにした方が良いと思ったが、現状は知っていても良さそう（YAGNIになりそう）なので保留
・WebSocketSetEventsという名前は改善の余地がありそう

## その他
聞きたいこと
・こんな感じでWebSocketやRPC引き離してみたが、問題は無いでしょうか？
（そもそもこのリファクタの方向性で大丈夫か？）

close #30 
close #33

kmurayama-aim committed:
[リファクタリング] MainControllerからWebSocketSharpの参照を取り除く

 MainControllerの中でWebSocketSharpを使っていた部分をWebSocketInitializerに隔離
 MainControllerがWebSocketInitializerをAwakeで見つけないとエラーが出る

kmurayama-aim committed:
WebSocketのアドレスが間違っていたため修正

kmurayama-aim committed:
PackagesUpdate

kmurayama-aim committed:
[リファクタリング] WebSocektSample.Rpcのデータを受け取るクラスを作成

次のリファクタでこれらクライアント側のクラスにWebSocektSample.Rpcのデータを入れる
現状クライアントのコードはRPCに依存しているので、これらのクラスに依存をさせたい

kmurayama-aim committed:
[リファクタリング] MainControllerからWebSocektSample.Rpcの参照を取り除く

WebSocketSample.Rpcのデータをクライアントのクラスに変換。
WebSocketSample.RpcはWebSocketInitializerに隔離し、
MainControllerのドメイン層への依存を削除

kmurayama-aim committed:
[リファクタリング] WebSocketInitializerがWebSocketの初期化のみを行うように変更

WebSocketInitializerはWebSocketの初期化についてのみで変更を行いたい
そのためWebSocketSetEventsを新たに作成しMainControllerの参照を任せた
また送られてきたデータについてもWebSocketInitializerで
処理方法を知ったままではいけないと思い、WebSocketSetEventsで処理するようにした

fhiraon commented on Client/Assets/Scripts/WebSocketSetEvents.cs:
```diff
@@ -0,0 +1,86 @@
+    {
+        mainControllerSc.Login();
+    }
+}
```
サーバー用の型をクライアント用の型に一度ラップしているのは良いと思います

fhiraon commented on Client/Assets/Scripts/DeleteItem.cs:
```diff
@@ -0,0 +1,13 @@
+
+public class DeleteItem
+{
+    public int Id;
```
どこでも変更できるようにしてあると意図せずにデータを壊してしまうので、できる限り外から書き換えることを許さない方がいいです

fhiraon commented on Client/Assets/Scripts/Environment.cs:
```diff
@@ -0,0 +1,12 @@
+
+public class Environment
+{
+    public List<Item> Items;
```
List を公開すると外から Add や AddRange, Remove とかされてしまうので、 IEnumerable にして外から操作できないようにするのが安全

fhiraon commented on Client/Assets/Scripts/WebSocketInitializer.cs:
```diff
@@ -0,0 +1,124 @@
+        var jsonMessage = JsonUtility.ToJson(getItemRpc);
+        return jsonMessage;
+    }
+}
```
特定のパラメータから通信用のデータを作成するのはちょっとやりすぎかな
Json を用意するのは別クラスにして責任を分けたほうが良いかな
あと、変に Unity意識して MonoBehaviour を継承する必要はないので、Unity に依存する必要がないクラスは極力継承しないようにするとよさそう

kmurayama-aim committed:
[リファクタリング] ラップ用クラスにおける変数のアクセスレベルを変更

レビューを取り入れ変更
https://github.com/kmurayama-aim/senka-hal-tokyo/pull/32#discussion_r245581847
Publicな変数となっており外部からの変更ができたため、読み取り専用のgetterに変更

kmurayama-aim committed:
[リファクタリング] Json文字列への変換処理をJsonMessageCreaterに任せた

レビューを取り込み変更
https://github.com/kmurayama-aim/senka-hal-tokyo/pull/32#discussion_r245585207
WebSocketInitializerがJson文字列への変換処理を行っていたが、
初期化以外のことをやるべきでは無いため変更

kmurayama-aim committed:
[リファクタリング] 外部に公開していたListをIEnumerableに変更

レビューを取り込み変更
https://github.com/kmurayama-aim/senka-hal-tokyo/pull/32#discussion_r245582891
Listは外部で変更が可能なため危険

kmurayama-aim committed:
[リファクタリング] MainControllerのJsonやWebSocketの情報を削除した

MainControllerがサーバーへメッセージを送る際、JsonやWebSocketの情報を知らなくて済むようになった

kmurayama-aim committed:
[リファクタリング] Logoutの方が誤解が少ないと思い変更

kmurayama-aim committed:
[リファクタリング] MainThreadExecutorへ登録する必要がなかったため削除

WebSocketSetEventsのメソッド側でMainThreadExecutorへの登録を行っていたため二回登録していた

kmurayama-aim committed:
[リファクタリング] ログインはMainController側で任意のタイミングで行いたいので変更

そもそも初期化とは関係なく、Monobehaviourも不要になる

kmurayama-aim committed:
[リファクタリング] Monobehaviourを継承する必要のないクラスから取り除いた

レビューを取り込み変更
https://github.com/kmurayama-aim/senka-hal-tokyo/pull/32#discussion_r245585207

kmurayama-aim committed:
[リファクタリング] デメテルの法則に基づき変更

kmurayama-aim committed:
[リファクタリング] 別のデータに整形するクラスなので、Formatの方が正しいと判断したため変更

参考記事：
https://qiita.com/KeithYokoma/items/ee21fec6a3ebb5d1e9a8#%E3%83%87%E3%83%BC%E3%82%BF%E3%82%92%E5%8A%A0%E5%B7%A5%E3%81%99%E3%82%8B%E3%83%AC%E3%82%A4%E3%83%A4
参考部分のGyazo：
https://gyazo.aiming-inc.biz/data/0/7/071976c077007eac20e19a39336bb425.png

kmurayama-aim committed:
[リファクタリング] 受け取ったJson文字列をデシリアライズするクラスを作成

サーべーへの送信時にJson文字列を作るクラスがあったものの、
受け取り時にデシリアライズするクラスが無かったため作成

kmurayama-aim committed:
[リファクタリング] メソッドを用意したにもかかわらず変更し忘れていた部分の修正

kmurayama-aim committed:
[リファクタリング] メソッド同士の間に１行間を開けた

Senkaのコードでは一貫して１行空いていたため

kmurayama-aim committed:
[リファクタリング] Jsonのデシリアライズ処理を任せ忘れていた部分の修正

kmurayama-aim committed:
[リファクタリング] WebSocketSetEventsの初期化がメソッドに依存しないように変更
